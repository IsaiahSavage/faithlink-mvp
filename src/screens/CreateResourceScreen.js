import React, { useState } from 'react';
import {
  ActivityIndicator,
  KeyboardAvoidingView,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import { Button, TextInput } from 'react-native-paper';
import { toProperCase } from '../utils/StringUtils';
import { useUserContext } from '../contexts/UserContext';
import { FIRESTORE_DB } from '../../firebase/firebaseConfig';
import { Timestamp, collection, addDoc, doc } from 'firebase/firestore';

const CreateResourceScreen = () => {
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');
  const [tags, setTags] = useState('');
  const [loading, setLoading] = useState(false);

  const { state } = useUserContext();

  // TODO: in the future, implement more styling, customization for the HTML content
  const constructHtml = (title, content) => {
    return `<div style='margin: 0 2.5rem; text-align: center;'><h1>${toProperCase(
      title,
    )}</h1><p>${content}</p></div>`;
  };

  /**
   * Submits the form data to create a new resource.
   * @async
   * @function submit
   * @returns {Promise<void>} A promise that resolves when the submission is complete.
   */
  const submit = async () => {
    setLoading(true);
    if (!title || !content) {
      alert('Please fill out all required fields.');
      setLoading(false);
      return;
    }

    const html = constructHtml(title, content);

    try {
      const today = new Date(Date.now());
      const timestamp = Timestamp.fromDate(today);

      const docRef = await addDoc(collection(FIRESTORE_DB, 'resources'), {
        author: doc(FIRESTORE_DB, `users/${state.userID}`),
        createdOn: timestamp,
        group: doc(FIRESTORE_DB, `groups/${state.userData.groupID}`),
        mediaType: 'article',
        source: {
          html: html,
        },
        tags: toProperCase(tags)
          .split(',')
          .map((tag) => tag.trim()),
        title: toProperCase(title).split(' '),
      });
    } catch (error) {
      console.log('Error adding document: ', error);
      alert('Error adding document: ' + error.message || 'Unknown error');
    } finally {
      setLoading(false);
    }
  };

  return (
    <ScrollView>
      <KeyboardAvoidingView
        style={styles.container}
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      >
        <Text style={styles.title}>Create Resource</Text>
        <View style={styles.inputContainer}>
          <TextInput
            style={[styles.input, styles.required]}
            placeholder="Title"
            value={title}
            onChangeText={setTitle}
            autoCapitalize="words"
            enablesReturnKeyAutomatically={true}
            maxLength={100}
            theme={{ colors: { primary: '#002857' } }}
          />
          <TextInput
            style={[styles.input, styles.required]}
            placeholder="Content"
            value={content}
            onChangeText={setContent}
            enablesReturnKeyAutomatically={true}
            maxLength={5000}
            numberOfLines={4}
            multiline={true}
            theme={{ colors: { primary: '#002857' } }}
          />
          <TextInput
            style={styles.input}
            placeholder="Tags (comma separated)"
            value={tags}
            onChangeText={setTags}
            autoCapitalize="words"
            maxLength={5000}
            theme={{ colors: { primary: '#002857' } }}
          />
          {loading ? (
            <ActivityIndicator size={'large'} color={'#002857'} />
          ) : (
            <Button
              mode="contained"
              buttonColor="#002857"
              style={styles.btnContainer}
              labelStyle={styles.submitText}
              dark={true}
              onPress={submit}
            >
              Create
            </Button>
          )}
        </View>
      </KeyboardAvoidingView>
    </ScrollView>
  );
};

export default CreateResourceScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
  title: {
    color: '#002857',
    fontSize: 30,
    marginTop: 20,
    marginBottom: 20,
  },
  inputContainer: {
    justifyContent: 'center',
  },
  input: {
    fontSize: 15,
    minWidth: 300,
    marginBottom: 20,
    backgroundColor: '#E8E8E8',
  },
  btnContainer: {
    alignSelf: 'center',
    width: '80%',
    alignItems: 'center',
    borderRadius: 100,
    marginTop: 20,
  },
  submitText: {
    fontSize: 20,
    width: '100%',
    height: 'auto',
  },
  required: {
    fontWeight: 'bold',
  },
});
