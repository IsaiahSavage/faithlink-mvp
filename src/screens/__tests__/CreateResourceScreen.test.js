import React from 'react';
import { render } from '@testing-library/react-native';

import CreateResourceScreen from '../CreateResourceScreen';
/*
jest.mock('../../contexts/UserContext.js', () => ({
    __esModule: true,
    useUserContext: jest.fn(() => ({
      state: {
        // Mocked user data
        userID: 1,
        userData: {
          id: 1,
          first: 'John',
          last: 'Doe',
          email: 'john.doe@example.com',
        },
      },
      // Mocked function
      dispatch: jest.fn(),
    })),
  })); */

describe('CreateResource Screen', () =>{

    //page renders
    it('page renders correctly', () =>{
        render(<CreateResourceScreen/>); 
     });
 
     //is not changing between loads
     it('matches the snapshot', () => {
         const { toJSON } = render(<CreateResourceScreen />);
         expect(toJSON()).toMatchSnapshot();
       });
 
     //title matches
     it('text is the same', () =>{
         const {getByText} = render(<CreateResourceScreen/>);
         const title = getByText(/Create Resource/);
         expect(title).toBeTruthy();
     });
});