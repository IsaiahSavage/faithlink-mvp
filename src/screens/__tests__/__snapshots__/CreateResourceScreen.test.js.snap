// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateResourceScreen matches the snapshot 1`] = `
<RCTScrollView>
  <View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "flex": 1,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <Text
        style={
          {
            "color": "#002857",
            "fontSize": 30,
            "marginBottom": 20,
            "marginTop": 20,
          }
        }
      >
        Create Resource
      </Text>
      <View
        style={
          {
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "rgba(231, 224, 236, 1)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              {
                "backgroundColor": "#E8E8E8",
                "marginBottom": 20,
                "minWidth": 300,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(73, 69, 79, 1)",
                "bottom": 0,
                "height": 1,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": [
                  {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
            testID="text-input-underline"
          />
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "minHeight": 56,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="words"
              cursorColor="#002857"
              editable={true}
              enablesReturnKeyAutomatically={true}
              maxFontSizeMultiplier={1.5}
              maxLength={100}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Title"
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="#002857"
              style={
                [
                  {
                    "margin": 0,
                  },
                  {
                    "height": 56,
                  },
                  {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "minWidth": 65,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  [
                    {},
                  ],
                  undefined,
                ]
              }
              testID="text-input-flat"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "backgroundColor": "rgba(231, 224, 236, 1)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              {
                "backgroundColor": "#E8E8E8",
                "marginBottom": 20,
                "minWidth": 300,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(73, 69, 79, 1)",
                "bottom": 0,
                "height": 1,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": [
                  {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
            testID="text-input-underline"
          />
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "minHeight": 56,
                },
              ]
            }
          >
            <TextInput
              cursorColor="#002857"
              editable={true}
              enablesReturnKeyAutomatically={true}
              maxFontSizeMultiplier={1.5}
              maxLength={5000}
              multiline={true}
              numberOfLines={4}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Content"
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="#002857"
              style={
                [
                  {
                    "margin": 0,
                  },
                  {},
                  {
                    "paddingBottom": 23,
                    "paddingTop": 23,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "minWidth": 65,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "textAlign": "left",
                    "textAlignVertical": "top",
                  },
                  false,
                  [
                    {},
                  ],
                  undefined,
                ]
              }
              testID="text-input-flat"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "backgroundColor": "rgba(231, 224, 236, 1)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              {
                "backgroundColor": "#E8E8E8",
                "marginBottom": 20,
                "minWidth": 300,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(73, 69, 79, 1)",
                "bottom": 0,
                "height": 1,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": [
                  {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
            testID="text-input-underline"
          />
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "minHeight": 56,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="words"
              cursorColor="#002857"
              editable={true}
              maxFontSizeMultiplier={1.5}
              maxLength={5000}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Tags (comma separated)"
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="#002857"
              style={
                [
                  {
                    "margin": 0,
                  },
                  {
                    "height": 56,
                  },
                  {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": undefined,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "minWidth": 65,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  [
                    {},
                  ],
                  undefined,
                ]
              }
              testID="text-input-flat"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          collapsable={false}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#002857",
              "borderRadius": 100,
              "marginTop": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": "80%",
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#002857",
                "borderColor": "transparent",
                "borderRadius": 100,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 100,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          {
                            "color": "#ffffff",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "fontSize": 20,
                            "height": "auto",
                            "width": "100%",
                          },
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Create
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
